### PR TITLE
[FLINK-17868][table-planner-blink] Fix proctime in DDL can not work in batch mode

### PR DESCRIPTION
**Pull request by LadyForest**
## What is the purpose of the change

* This pull request tries to fix the issue that `PROCTIME` generates null value in batch mode for the blink planner, by introducing `FlinkRelTimeIndicatorProgram ` to `FlinkBatchProgram` (*i.e.* reuse `RelTimeIndicatorConverter`) to rewrite `PROCTIME()` to `PROCTIME_MATERIALIZE(PROCTIME())`, which is consistent with stream mode.  More discussion on the implementation details can be founded at [FLINK-17868](https://issues.apache.org/jira/browse/FLINK-17868).

## Brief change log

  - *The changes applied to `FlinkBatchProgram`, `FlinkOptimizationContext`, `FlinkRelTimeIndicatorProgram`, `StreamOptimizationContext`, `BatchCommonSubGraphBasedOptimizer` is to reuse `RelTimeIndicatorConverter` under batch mode.*
  - *The changes applied to `TableStreamOperator`, `StreamExecCalc`, `StreamExecCorrelate`, `StreamExecDataStreamScan`, `StreamLegacyTableSourceScan` is to merge `AbstractProcessStreamOperator` into `TableStreamOperator`. As a result, `TableStreamOperator` will be the common base class for streaming and batch operators to share.*
  - *The changes applied to `BatchLogicalWindowAggregateRule#getOutAggregateGroupExpression` is to avoid `RelOptUtil#verifyTypeEquivalence` failing, because after introducing `RelTimeIndicatorConverter`, the time attribute in window expression is modfied to `TIMESTAMP`.*

## Verifying this change

This change added tests and can be verified as follows:

  - *`TableScanTest#testDDLWithProctime` and `TableSourceTest#testSimpleProjectWithProctime` is to verify the optimized plan.*
  - *`TableScanITCase#testProctimeTableSource` and `TableSourceITCase#testSimpleProjectWithProcTime` is to verify the execution result.*
  - *The changes made to `(Non)BufferDataOverWindowOperatorTest` adapts to the change of `TableStreamOperator`.*
 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by LadyForest** [comment]
Hi @wuchong, would you mind helping to review this PR when you're free?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkBatchProgram.scala [hunk 1]
```diff
@@ -73,6 +74,8 @@ object FlinkBatchProgram {
           .build(), "convert table references after sub-queries removed")
         .build()
     )
+    // convert time indicators
+    chainedProgram.addLast(TIME_INDICATOR, new FlinkRelTimeIndicatorProgram)
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkBatchProgram.scala:78]: We put TIME_INDICATOR phase under decorrelate in streaming programs, I think we should keep the same order in batch program to put the phase after decorrelate.

**Review by wuchong — COMMENTED** [review]
Thanks for the contribution @LadyForest , the changes looks good to in general. I only left some minor comments. 

Besides, I think we can also remove `AbstractProcessStreamOperator` class now. It is not used anymore.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkBatchProgram.scala (hunk 1 above)
Comment (LadyForest) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkBatchProgram.scala:78]: Good catch!

**Comment by LadyForest** [comment]
@flinkbot run azure

**Comment by LadyForest** [comment]
Hi @wuchong, I addressed the comments as you suggested, and please take a look when you're free. 

Thanks for helping polish this PR.

Best, Jane

**Review by wuchong — APPROVED** [review]
Thanks for the great work @LadyForest . 

LGTM.

**Pull request merged**